support-workers/src/main/scala/com/gu/zuora/subscriptionBuilders/DigitalSubscriptionBuilder.scala

```diff
@@ -5,90 +5,166 @@ import java.util.UUID
+
   def build(
     digitalPack: DigitalPack,
     requestId: UUID,
     subscriptionPaymentType: SubscriptionPaymentType,
     environment: TouchPointEnvironment,
     today: () => LocalDate
-  )(implicit ec: ExecutionContext): EitherT[Future, Either[PromoError, GetCodeStatus.RedemptionInvalid], SubscriptionData] = {
+  )(implicit ec: ExecutionContext): BuildResult = {
+    val productRatePlanId = validateRatePlan(digitalPack.productRatePlan(environment, digitalPack.readerType), digitalPack.describe)
 
-    val contractEffectiveDate = today()
-    val delay = subscriptionPaymentType match {
-      case direct: SubscriptionPaymentDirect => direct.zuoraDigitalPackConfig.defaultFreeTrialPeriod + direct.zuoraDigitalPackConfig.paymentGracePeriod
-      case _: SubscriptionPaymentCorporate => 0
+    subscriptionPaymentType match {
+      case purchase: SubscriptionPurchase =>
+        buildPurchase(today(), productRatePlanId, requestId, digitalPack.readerType, purchase)
+      case redemption: SubscriptionRedemption =>
+        buildRedemption(today(), productRatePlanId, requestId, digitalPack.readerType, redemption)
     }
-    val contractAcceptanceDate = contractEffectiveDate.plusDays(delay)
+  }
 
-    val productRatePlanId = validateRatePlan(digitalPack.productRatePlan(environment, digitalPack.readerType), digitalPack.describe)
+  def buildPurchase(
+    today: LocalDate,
+    productRatePlanId: ProductRatePlanId,
+    requestId: UUID,
+    readerType: ReaderType,
+    purchase: SubscriptionPurchase
+  )(implicit ec: ExecutionContext): BuildResult = {
+
+    val (contractAcceptanceDelay, autoRenew, initialTerm) = if (readerType == Gift)
+      (0, false, purchase.billingPeriod.monthsInPeriod)
+    else
+      (purchase.config.defaultFreeTrialPeriod + purchase.config.paymentGracePeriod, true, 12)
+
+    val contractAcceptanceDate = today.plusDays(contractAcceptanceDelay)
 
     val subscriptionData = buildProductSubscription(
       requestId,
       productRatePlanId,
       contractAcceptanceDate = contractAcceptanceDate,
-      contractEffectiveDate = contractEffectiveDate,
-      readerType = digitalPack.readerType
+      contractEffectiveDate = today,
+      readerType = readerType,
+      autoRenew = autoRenew,
+      initialTerm = initialTerm
     )
 
-    subscriptionPaymentType match {
-      case SubscriptionPaymentDirect(_, maybePromoCode, country, promotionService) =>
-        EitherT.fromEither[Future](
-          applyPromoCode(promotionService, maybePromoCode, country, productRatePlanId, subscriptionData)
-            .left.map(Left.apply)
-        )
-      case SubscriptionPaymentCorporate(redemptionData, getCodeStatus) =>
-        withRedemption(subscriptionData.subscription, redemptionData, getCodeStatus)
-          .map(subscription => subscriptionData.copy(subscription = subscription))
-          .leftMap(Right.apply)
+    val withRedemptionCode = generateRedemptionCode(readerType, purchase.billingPeriod, subscriptionData)
```

Comment:
How about `addRedemptionCodeIfGift` (and `applyPromoIfAvailable`)?
I'm already not totally happy with the existing naming, this and the applyPromoCode when I've done previous work on this code.
It would be ok if it was `maybePromoCode.map(applyPromoCode(...))` and if the `if (readerType == Gift)` was outside.  But both of these are actually conditional within the logic.
`generateRedemptionCode` is not just conditionally getting the code generated but it's actually adding it to the subscription data.